### PR TITLE
EVMCC & Fab3 Dependency Split

### PR DESCRIPTION
**Pull request by swetharepakula**
* [x] Ran the basic checks: `make basic-checks`
* [x] Ran the unit tests: `make unit-test`
* [x] Ran the integration tests: `make integration-test`
* [x] Added the associated JIRA ticket number in the commit message title
* [x] Link to the JIRA Ticket: 
https://jira.hyperledger.org/browse/FAB-17053
https://jira.hyperledger.org/browse/FAB-16919

### CODE REVIEW

File: gotools.mk [hunk 1]
```diff
@@ -22,19 +22,14 @@ gotools-clean:
 	-@rm -rf $(BUILD_DIR)/gotools
 
 # Special override for ginkgo since we want to use the version vendored with the project
+gotool.dep: GINKGO_VERSION ?= "v1.10.2"
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply gotools.mk]: Regarding the version, I feel like we were tied to the version in fabric, previously, so that's pretty cool if it's not anymore.

File: fab3/go.mod [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	github.com/pkg/errors v0.8.1
+	github.com/spf13/cobra v0.0.1
+	github.com/spf13/viper v1.4.0
+	github.com/zmap/zlint v1.0.2 // indirect
+	go.uber.org/zap v1.10.0
+)
+
+replace github.com/hyperledger/fabric-chaincode-evm/evmcc => ../evmcc
+
+replace github.com/go-kit/kit => github.com/go-kit/kit v0.8.0
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply fab3/go.mod]: yes this is like the override from dep. It is isn't the only way, but in this situation it is needed because i think burrow and fabric use go-kit and since burrow explicitly asks for a newer version dep doesn't let you require a lower version, so we need to use the replace syntax.

File: Makefile [hunk 3]
```diff
@@ -54,22 +54,18 @@ gotools: gotools-install
...
-linter: gotool.goimports check-deps
+linter: gotool.goimports
 	@echo "LINT: Running code checks.."
 	@scripts/golinter.sh
 
-check-deps: gotool.dep
-	@echo "DEP: Checking for dependency issues.."
-	dep version
-	dep check
-
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply Makefile]: i like that, have to see how that works with vendor for evmcc and integration

File: Makefile [hunk 4]
```diff
@@ -54,22 +54,18 @@ gotools: gotools-install
 
 unit-test: gotool.ginkgo
 	@echo "Running unit-tests"
-	ginkgo -p -randomizeAllSpecs -randomizeSuites -requireSuite -noColor -keepGoing -race -skipPackage integration -r
+	ginkgo -p -randomizeAllSpecs -randomizeSuites -requireSuite -noColor -keepGoing -race -r evmcc
+	cd fab3 && GO111MODULE=on ginkgo -p -randomizeAllSpecs -randomizeSuites -requireSuite -noColor -keepGoing -race -r
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply Makefile]: I wasn't happy about this either and hopefully as tools improve and can deal with mixture of modules and not this will simplify

File: evmcc/go.sum [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7/go.mod h1:dt/ZhP58zS4L8KSrWDmTeBkI65Dw0HsyUHuEVlX15mw=
+gopkg.in/yaml.v2 v2.0.0-20170812160011-eb3733d160e7/go.mod h1:JAlM8MvJe8wmxCU4Bli9HhUf9+ttbYbLASfIpnQbh74=
+gopkg.in/yaml.v2 v2.2.1 h1:mUhvW9EsL+naU5Q3cakzfE91YhliOondGd6ZrsDBHQE=
+gopkg.in/yaml.v2 v2.2.1/go.mod h1:hI93XBmqTisBFMUTm0b8Fm+jr3Dg1NNxqwp+5A1VGuI=
+gopkg.in/yaml.v2 v2.2.2 h1:ZCJp+EgiOT7lHqUV2J862kp8Qj64Jo6az82+3Td9dZw=
+gopkg.in/yaml.v2 v2.2.2/go.mod h1:hI93XBmqTisBFMUTm0b8Fm+jr3Dg1NNxqwp+5A1VGuI=
+<<<<<<< HEAD
+gotest.tools v2.1.0+incompatible h1:5USw7CrJBYKqjg9R7QlA6jzqZKEAtvW82aNmsxxGPxw=
+=======
+>>>>>>> go-modules-take-2
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply evmcc/go.sum]: fixed

File: ci/azure-pipelines.yml [hunk 6]
```diff
@@ -11,7 +11,7 @@ pr:
 variables:
   GOPATH: $(Agent.BuildDirectory)/go
   PATH: $(Agent.BuildDirectory)/go/bin:/bin:/usr/bin:/sbin:/usr/sbin:/usr/local/bin:/usr/local/sbin
-  GO_VER: 1.10.4
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply ci/azure-pipelines.yml]: im okay doing 13, but we should probably test against a range of go versions then. should i create a new jira for that?

File: gotools.mk [hunk 7]
```diff
@@ -22,19 +22,14 @@ gotools-clean:
 	-@rm -rf $(BUILD_DIR)/gotools
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply gotools.mk]: good catch. should be fixed now

File: gotools.mk [hunk 8]
```diff
@@ -22,19 +22,14 @@ gotools-clean:
...
-gotool.dep: DEP_VERSION ?= "v0.5.4"
-gotool.dep:
-	@GOPATH=$(abspath $(GOTOOLS_GOPATH)) go get -d -u github.com/golang/dep
-	@git -C $(abspath $(GOTOOLS_GOPATH))/src/github.com/golang/dep checkout -q $(DEP_VERSION)
-	@echo "Building github.com/golang/dep $(DEP_VERSION) -> dep"
-	@GOPATH=$(abspath $(GOTOOLS_GOPATH)) GOBIN=$(abspath $(GOTOOLS_BINDIR)) go install -ldflags="-X main.version=$(DEP_VERSION) -X main.buildDate=$$(date '+%Y-%m-%d')" github.com/golang/dep/cmd/dep
+	@GOPATH=$(abspath $(GOTOOLS_GOPATH)) go get -d -u github.com/onsi/ginkgo
+	@git -C $(abspath $(GOTOOLS_GOPATH))/src/github.com/onsi/ginkgo checkout -q $(DEP_VERSION)
+	@echo "Building github.com/onsi/ginkgo/ginkgo $(GINKGO_VERSION)-> ginkgo"
+	@GOPATH=$(abspath $(GOTOOLS_GOPATH)) GOBIN=$(abspath $(GOTOOLS_BINDIR)) go install -ldflags="-X main.version=$(DEP_VERSION) -X main.buildDate=$$(date '+%Y-%m-%d')" github.com/onsi/ginkgo/ginkgo
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply gotools.mk]: fixed

File: evmcc/vendor/modules.txt [hunk 9]
```diff
@@ -0,0 +1,337 @@
+# github.com/Azure/go-ansiterm v0.0.0-20170929234023-d6e3b3328b78
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply evmcc/vendor/modules.txt]: go mod I don't think is as good as dep at only pulling in the files needed. it pulls in entire projects which means we probably get more dependencies than before.

File: Makefile [hunk 10]
```diff
@@ -96,7 +92,7 @@ fab3: bin/fab3
 .PHONY: bin/fab3 # let 'go build' handle caching and whether to rebuild
 bin/fab3:
 	mkdir -p bin/
-	go build -o bin/fab3 ./fab3/cmd
+	cd fab3 && GO111MODULE=on go build -o ./../bin/fab3 ./cmd
```
Comment (MHBauer) [inline Makefile]: someone showed me an interesting GOBIN trick, but it can wait.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (MHBauer) [reply Makefile]: I agree, but it may require ginkgo to become module aware.

File: Gopkg.toml [hunk 11]
```diff
@@ -1,112 +0,0 @@
...
-  revision = "505ab145d0a99da450461ae2c1a9f6cd10d1f447"
-  name = "golang.org/x/crypto"
-
-[[override]]
-  name = "github.com/btcsuite/btcd"
-  revision = "ed77733ec07dfc8a513741138419b8d9d3de9d2d"
-
-[[override]]
-  name = "github.com/go-kit/kit"
-  version = "0.8.0"
```
Comment (MHBauer) [inline Gopkg.toml]: this being the version of go-kit we had before this change.

File: integration/go.mod [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+	github.com/onsi/ginkgo v1.10.2
+	github.com/onsi/gomega v1.7.0
+	github.com/tedsuo/ifrit v0.0.0-20191009134036-9a97d0632f00
+	github.com/willf/bitset v1.1.10 // indirect
+	go.etcd.io/etcd v3.3.17+incompatible // indirect
+)
+
+replace github.com/go-kit/kit => github.com/go-kit/kit v0.7.0
+
+replace github.com/hyperledger/burrow => github.com/hyperledger/burrow v0.24.4
```
Comment (MHBauer) [inline integration/go.mod]: I think these two replaces exist due to the v0.x versions might choose non-compiling things otherwise

File: integration/go.mod [hunk 13]
```diff
@@ -0,0 +1,24 @@
+module github.com/hyperledger/fabric-chaincode-evm/integration
+
+replace github.com/hyperledger/fabric-chaincode-evm/fab3 => ../fab3
+
+replace github.com/hyperledger/fabric-chaincode-evm/evmcc => ../evmcc
```
Comment (MHBauer) [inline integration/go.mod]: I was going to ask if all the replaces should be together, but I see that these are the local refs to the other repos.

File: fab3/go.mod [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+	github.com/onsi/ginkgo v1.10.2
+	github.com/onsi/gomega v1.7.0
+	github.com/pkg/errors v0.8.1
+	github.com/spf13/cobra v0.0.1
+	github.com/spf13/viper v1.4.0
+	github.com/zmap/zlint v1.0.2 // indirect
+	go.uber.org/zap v1.10.0
+)
+
+replace github.com/hyperledger/fabric-chaincode-evm/evmcc => ../evmcc
```
Comment (MHBauer) [inline fab3/go.mod]: here the local ref replace is below, maybe we should be consistent.

File: scripts/check_license.sh [hunk 15]
```diff
@@ -15,7 +15,7 @@ function filterExcludedFiles {
   CHECK=`echo "$CHECK" \
 		| grep -v "^\.git/" \
 		| grep -v "^\.build/" \
-		| grep -v "^vendor/" \
+		| grep -v "vendor/" \
```
Comment (MHBauer) [inline scripts/check_license.sh]: this changes the irgnore from toplevel to any level.

File: scripts/golinter.sh [hunk 16]
```diff
@@ -4,18 +4,24 @@
...
+"./integration/fab3"
+"./integration/helpers"
 "./fab3"
-"./integration"
-"./statemanager"
 )
 
-for i in "${arr[@]}"
+declare -a goModules=(
+)
```
Comment (MHBauer) [inline scripts/golinter.sh]: this is probably used in a later commit?

File: scripts/golinter.sh [hunk 17]
```diff
@@ -13,10 +13,10 @@ declare -a vendoredModules=(
 "./integration/e2e"
 "./integration/fab3"
 "./integration/helpers"
-"./fab3"
 )
 
 declare -a goModules=(
+"./fab3"
 )
```
Comment (MHBauer) [inline scripts/golinter.sh]: Yes, I see it now, 2nd commit.

File: scripts/golinter.sh [hunk 18]
```diff
@@ -31,6 +31,18 @@ do
...
+    echo ">>>Checking $i with goimports"
+    OUTPUT="$(goimports -l ./$i || true )"
+    if [[ $OUTPUT ]]; then
+        echo "The following files contain goimports errors"
+        echo $OUTPUT
+        echo "The goimports command 'goimports -l -w' must be run for these files"
+        exit 1
+    fi
+done
+
```
Comment (MHBauer) [inline scripts/golinter.sh]: something something shell magic to merge these two arrays and remove the duplication.

needs research, I don't know off the top of my head.

File: scripts/golinter.sh [hunk 19]
```diff
@@ -41,3 +53,18 @@ do
...
     fi
 done
+
+for i in "${goModules[@]}"
+do
+    echo ">>>Checking $i with go vet"
+    pushd $i
+        OUTPUT="$(GO111MODULE=on go vet .)"
+        exitCode=$?
+    popd
```
Comment (MHBauer) [inline scripts/golinter.sh]: this moves in and out of each listed module directory to run the vet command.

I wonder if the original invocations are compatible in the same way, to again merge the loops.

vet could probably be reduced to a ./... but not the other commands, so in the effort of sharing the same lists, this is fine. 🤷‍♂ 🤷‍♀

File: scripts/goListFiles.sh [hunk 20]
```diff
@@ -1,30 +0,0 @@
...
-    do
-        importpath=$GOPATH/src/$dep
-        if [ -d $importpath ];
-        then
-            find_golang_src $importpath
-        fi
-    done
-}
-
-list_deps $1 | sort | uniq
```
Comment (MHBauer) [inline scripts/goListFiles.sh]: it's separated out now, but this should probably be some separate pr or jira.

File: integration/go.mod [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+	github.com/hyperledger/fabric v1.4.0
+	github.com/hyperledger/fabric-chaincode-evm/fab3 v0.0.0
+	github.com/onsi/ginkgo v1.10.2
+	github.com/onsi/gomega v1.7.0
+	github.com/tedsuo/ifrit v0.0.0-20191009134036-9a97d0632f00
+	github.com/willf/bitset v1.1.10 // indirect
+	go.etcd.io/etcd v3.3.17+incompatible // indirect
+)
+
+replace github.com/go-kit/kit => github.com/go-kit/kit v0.7.0
```
Comment (MHBauer) [inline integration/go.mod]: should we force this to 0.8.0?

Comment (swetharepakula) [reply integration/go.mod]: Nope, we need to force to v0.7.0 because of fabric. We don't have a problem in evmcc and fab3 because we aren't building fabric in those situations and we don't have an compile issues there. In the integration tests we have to build Fabric 1.4.0 which is locked on v0.7.0 and v0.8.0 introduces compile errors.

File: scripts/goListFiles.sh (hunk 20 above)
Comment (swetharepakula) [reply scripts/goListFiles.sh]: It is a separate commit and has a separate jira. If it was a huge deletion would have made it into its own PR. But if you want I can separate it into its own PR. I found this in the process of making all the scripts work with go modules

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (swetharepakula) [reply Makefile]: ginkgo becoming module aware may be available in the close future, but that is not the issue. go test can't selectively use go modules. I can't remember, but I don't think go test can handle dealing with multiple modules even if all are using go modules.

File: fab3/go.mod (hunk 14 above)
Comment (swetharepakula) [reply fab3/go.mod]: This happened because of how go mod edit works. I will go through and manually fix them.

File: integration/go.mod (hunk 12 above)
Comment (swetharepakula) [reply integration/go.mod]: Yup that is part of the need for these replaces. The other is that these are requirements of fab3 so we need to put them as overrides because from go's perspective integration has no dependency because it builds fab3 and doesn't actually use fab3 as a library.

File: scripts/goListFiles.sh (hunk 20 above)
Comment (MHBauer) [reply scripts/goListFiles.sh]: nah, no big deal. it's here and looks fine.

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (swetharepakula) [reply Makefile]: Is it okay to move this to another pr? Or do you want this addressed in this one?

File: integration/go.mod (hunk 13 above)
Comment (swetharepakula) [reply integration/go.mod]: moved this down to both so all replaces are together.

File: scripts/golinter.sh (hunk 17 above)
Comment (swetharepakula) [reply scripts/golinter.sh]: yea not the cleanest division of the commits

File: integration/go.mod (hunk 13 above)
Comment (MHBauer) [reply integration/go.mod]: yay

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (MHBauer) [reply Makefile]: Yes, separate.

File: evmcc/vendor/modules.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (MHBauer) [reply evmcc/vendor/modules.txt]: makes sense for non-vendoring where a whole project is the smallest case, and no trimming is necessary because it's not stored in the repo, but is stored locally.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (MHBauer) [reply Makefile]: bleh.

**Comment by MHBauer** [comment]
I didn't delete anything. I thought we turned that off?

**Comment by swetharepakula** [comment]
No, I don't think we turned that off.
